### ISSUE TITLE
Consider throwing the error for uninitialized instance variables from *abstract* classes only in the child class

### ISSUE THREAD

**Issue by thomkeh** [microsoft/pyright#3990]
**Is your feature request related to a problem? Please describe.**
With `reportUninitializedInstanceVariable` set to `true`, _pyright_ will report instance variables that are declared but never initialized, which is a great feature! However, I feel like this should be allowed in abstract base classes, and the error should only be thrown in the child classes that don't initialize the variable. This is in fact the behavior of *pyre*.

Consider this code:
```python
from abc import ABC

class Sized(ABC):
    len: int             # pyright reports error on this line
    def __len__(self) -> int:
        return self.len

class A(Sized):
    def __init__(self) -> None:
        self.len = 10

class B(Sized):          # pyre reports error on this line
    def __init__(self) -> None:
        pass

print(len(A()))
print(len(B()))  # runtime error
```

_pyre_ gives the following error:
```
main.py:12:0 Uninitialized attribute [13]: Attribute `len` inherited from abstract class `Sized` in class `B` to have type `int` but is never initialized.
```

However, _pyright_ gives the following error:
```
/main.py:4:5 - error: Instance variable "len" is not initialized in the class body or __init__ method (reportUninitializedInstanceVariable)
```

I think pyre's behavior is more useful because it's not dangerous if you have an uninitialized instance variable in an abstract class because you can't instantiate it anyway (well, you can't instantiate it if it has an abstract method...).

**Describe the solution you'd like**
I would like to see _pyright_ behave like _pyre_.

**Comment by erictraut** [comment]
If you change `Sized` from an `ABC` to a `Protocol`, pyright does emit an error in the same location as pyre. This is consistent with mypy, which emits an error in this location for a `Protocol` but not for an `ABC`.

I'll need to think more about the implications for changing this behavior for an `ABC` base class. In the meantime, you could consider switching from `ABC` to `Protocol` in your code.

**Comment by erictraut** [comment]
Thinking about this more... 

I don't think pyre is correct to emit an error where it does. The problem is that `B` could be subclassed by another class that provides a value for `len`.

One could argue that the expression `B()` should generate an error because `len` is not initialized, but `len` could be initialized in many different ways — many of them difficult for a static type checker to validate.

My conclusion is that pyright's current behavior here is correct — and that making the change that you've proposed will likely result in false positives.

If you want behavior that is closer to pyre's, you can switch from an `ABC` to a `Protocol`. If a class claims to derive from a protocol, it must initialize any instance variables. The same is not true for an `ABC` because a subclass could provide the initialization.
